ouroboros-consensus/src/Ouroboros/Consensus/Crypto/KES/Simple.hs

```diff
@@ -35,38 +38,79 @@ instance DSIGNAlgorithm d => KESAlgorithm (SimpleKES d) where
     newtype SigKES (SimpleKES d) = SigSimpleKES (SigDSIGN d)
         deriving Generic
 
+    encodeVerKeyKES = encode
+    encodeSignKeyKES = encode
+    encodeSigKES = encode
+
+    decodeSignKeyKES = decode
+    decodeVerKeyKES = decode
+    decodeSigKES = decode
+
     genKeyKES duration = do
         sks <- replicateM (fromIntegral duration) genKeyDSIGN
         let vks = map deriveVerKeyDSIGN sks
         return $ SignKeySimpleKES (vks, zip [0..] sks)
 
     deriveVerKeyKES (SignKeySimpleKES (vks, _)) = VerKeySimpleKES $ fromList vks
 
-    signKES j a (SignKeySimpleKES (vks, xs)) = case dropWhile (\(k, _) -> k < j) xs of
+    signKES toEnc j a (SignKeySimpleKES (vks, xs)) = case dropWhile (\(k, _) -> k < j) xs of
         []           -> return Nothing
         (_, sk) : ys -> do
-            sig <- signDSIGN a sk
+            sig <- signDSIGN toEnc a sk
             return $ Just (SigSimpleKES sig, SignKeySimpleKES (vks, ys))
 
-    verifyKES (VerKeySimpleKES vks) j a (SigSimpleKES sig) =
+    verifyKES toEnc (VerKeySimpleKES vks) j a (SigSimpleKES sig) =
         case vks !? fromIntegral j of
             Nothing -> False
-            Just vk -> verifyDSIGN vk a sig
+            Just vk -> verifyDSIGN toEnc vk a sig
 
 deriving instance DSIGNAlgorithm d => Show (VerKeyKES (SimpleKES d))
 deriving instance DSIGNAlgorithm d => Eq (VerKeyKES (SimpleKES d))
 deriving instance DSIGNAlgorithm d => Ord (VerKeyKES (SimpleKES d))
-deriving instance DSIGNAlgorithm d => Serialise (VerKeyKES (SimpleKES d))
+instance DSIGNAlgorithm d => Serialise (VerKeyKES (SimpleKES d)) where
+  encode (VerKeySimpleKES vvks) =
+       Enc.encodeListLen (fromIntegral $ Vec.length vvks)
+    <> Vec.foldl' (<>) mempty (fmap encodeVerKeyDSIGN vvks)
```

Comment:
Just wondering: why a strict fold here and not lazy (`foldr`)?

Reply:
No specific reason, I think it's just my default fold...

Reply:
Same here :slightly_smiling_face:, unless I think being lazy would be useful. I'm not sure, but maybe in this case it is? It depends on the length of the vector as well. How long is the length of this vector typically?

If it is very long, then being lazy could be beneficial, as we wouldn't need to keep the encoding of each element in memory and just get them on-demand. However, I think `Encoding` already is a `Builder`-like thing that allows efficient `mappend`s.

Reply:
Yes, `Encoding` is already such a thing.